nucliadb/nucliadb/common/datamanagers/labels.py

```diff
@@ -41,16 +43,87 @@ async def get_labels(self, kbid: str) -> kb_pb2.Labels:
     @classmethod
     async def inner_get_labels(cls, kbid: str, txn: Transaction) -> kb_pb2.Labels:
         labels = kb_pb2.Labels()
-        labels_key = KB_LABELS.format(kbid=kbid)
-        async for key in txn.keys(labels_key, count=-1):
-            labelset = await txn.get(key)
-            id = key.split("/")[-1]
-            if labelset:
-                ls = kb_pb2.LabelSet()
-                ls.ParseFromString(labelset)
-                labels.labelset[id].CopyFrom(ls)
+        labelset_ids = await cls._get_labelset_ids_bw_compat(kbid, txn)
+        for labelset_id in labelset_ids:
+            labelset = await txn.get(KB_LABELSET.format(kbid=kbid, id=labelset_id))
+            if not labelset:
+                continue
+            ls = kb_pb2.LabelSet()
+            ls.ParseFromString(labelset)
+            labels.labelset[labelset_id].CopyFrom(ls)
         return labels
 
+    @classmethod
+    async def _get_labelset_ids_bw_compat(
+        cls, kbid: str, txn: Transaction
+    ) -> list[str]:
+        labelsets = await cls._get_labelset_ids(kbid, txn)
+        if labelsets is not None:
+            return labelsets
+        # TODO: Remove this after migration #11
+        return await cls._legacy_get_labelset_ids(kbid, txn)
+
+    @classmethod
+    async def _legacy_get_labelset_ids(cls, kbid: str, txn: Transaction) -> list[str]:
+        labelsets = []
+        labels_key = KB_LABELS.format(kbid=kbid)
+        async for key in txn.keys(labels_key, count=-1, include_start=False):
+            lsid = key.split("/")[-1]
+            labelsets.append(lsid)
+        return labelsets
+
+    @classmethod
+    async def _get_labelset_ids(
+        cls, kbid: str, txn: Transaction
+    ) -> Optional[list[str]]:
+        key = KB_LABELSET_IDS.format(kbid=kbid)
+        data = await txn.get(key)
+        if not data:
+            return None
+        return orjson.loads(data)
+
+    @classmethod
+    async def _add_to_labelset_ids(
+        cls, kbid: str, txn: Transaction, labelsets: list[str]
+    ) -> None:
+        previous = await cls._get_labelset_ids(kbid, txn)
+        needs_set = False
+        if previous is None:
+            # TODO: Remove this after migration #11
+            needs_set = True
```

Comment:
This technically means on initial write, we're always doing a scan. I guess this is okay.

Reply:
yes, this will be like this until the migration has run. Afterwards, we will be able to remove the bw/compat code and skip the scan.